cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,256 @@
+    ));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(
+        target_addr,
+        notify_ready,
+        timeout,
+        rng_resp,
+    ));
+
+    tokio::select! {
+        task = validation_task => {
+            if let Err(err) = task.expect("Failed to run validator task") {
+                error!(error = %err, "Failed validation");
+                exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+            info!("Validation passed successfully...exiting");
+            Ok(())
+        }
+
+        _ = shutdown_tx.signaled() => {
+            if let Err(e) = server_task.await.expect("Failed to run outbound server task") {
+                error!(error = %e, "Failed to validate outbound routing configuration");
+            } else {
+                error!("Failed to validate due to server terminating early");
+            }
+
+            exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+#[tracing::instrument(level = "debug", skip_all)]
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready: Arc<Notify>,
+    timeout: time::Duration,
+    expected_resp: String,
+) -> Result<()> {
+    if tokio::time::timeout(timeout, ready.notified())
+        .await
+        .is_err()
+    {
+        anyhow::bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!("Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        debug_assert_eq!(target_addr, socket.peer_addr().unwrap());
+        debug!(original_dst = %target_addr, "Client connected to validation server");
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            is_readable.context("cannot read off client socket")?;
+        }
+
+        () = tokio::time::sleep(timeout) => {
+            anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
+
+    let mut buf = [0u8; 8];
+    let read_sz = stream.read_exact(&mut buf).await?;
+    let resp = String::from_utf8(buf[..expected_resp.len()].to_vec())?;
+    debug!(redirect_response = %resp, bytes_read = %read_sz);
+    anyhow::ensure!(
+        resp == expected_resp,
+        "expected client to receive {:?}, got {:?} instead",
+        expected_resp,
+        resp
+    );
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_server", skip(ready, shutdown))]
+async fn outbound_serve(
+    listen_addr: SocketAddr,
+    ready: Arc<Notify>,
+    shutdown: kubert::shutdown::Watch,
+    resp: String,
+) -> Result<()> {
+    let listener = TcpListener::bind(listen_addr)
+        .await
+        .expect("Failed to bind server");
+    info!("Listening for incoming connections");
+
+    ready.notify_one();
+
+    let accept_task = tokio::spawn(accept(listener, resp)).in_current_span();
+
+    let _handle = shutdown.signaled().await;
+    accept_task.inner().abort();
+    debug!("Received shutdown signal");
```

Comment:
nit, take it or leave it: it's not _really_ necessary to spawn the accept function here, since this task isn't really going to be doing anything else...alternatively, we could write something like this:
```suggestion
    tokio::select! {
        _ = accept(listener, resp) => unreachable!("`accept` function never returns"),
        _ = shutdown.signaled() => debug!("Received shutdown signal"),
    };
```